src/building-an-offchain-ipfs-app.md

```diff
@@ -1,5 +1,26 @@
-# Building an offchain::ipfs app
+# Building an `offchain::ipfs` app
 
-In this book we’ll cover the bare essentials on what you need to start building out your dApp
-with offchain::ipfs. We cover two Rust clients and one JavaScript client, but there are
-many other clients in other languages.
+In this section we’ll cover the essentials on what you'll need to start building out your
+application with `offchain::ipfs`. We do so by way of an example.
+
+1. First, we walk you through the example Substrate pallet that's included with `offchain::ipfs`
+2. Then, we provide example code from two Rust clients (`substrate-subxt` and
+`substrat-api-client`), and one JavaScript client (`polkadot.js`) as they relate
+
+We expect feedback on this pallet, but also we hope that the reference implementation will inspire
+builders to create their own pallets, expose their own JSON-RPC endpoints, and call them from
+their applications similarly.
+
+## How it all works
```

Comment:
Need more input here, both in terms of correctness and detail

Reply:
1. On chain initialisation or synced blocks, the Rust IPFS node is launched as a part of a offchain worker runtime and will stay running in the background.
2. User makes JSON-RPC call to submit an extrinsic to the node's runtime, using the callable functions exposed from the custom pallet.
1. The requests add to queues stored in Substrate storage database, defined in the custom pallet.
1. On block import (intervals depending on the type of request), the node's runtime passes some requests from the queues to a new offchain worker - let's call the request offchain worker
1. The request offchain worker performs future requests to the Rust IPFS node, and the returned results resolve the futures
1. The request offchain worker registers the result in its API the the node's runtime, which reads it and emits events declared in the custom pallet. The request offchain worker stops running